src/data-connect/data-connect-api-client-internal.ts

```diff
@@ -0,0 +1,233 @@
+        throw new FirebaseDataConnectError(
+          'invalid-argument', 'GraphqlOptions must be a non-null object');
+      }
+    }
+    const host = (process.env.DATA_CONNECT_EMULATOR_HOST || DATA_CONNECT_HOST);
+    const data = {
+      query,
+      ...(options?.variables && { variables: options?.variables }),
+    };
+    return this.getUrl(host, this.connectorConfig.location, this.connectorConfig.serviceId, endpoint)
+      .then(async (url) => {
+        const request: HttpRequestConfig = {
+          method: 'POST',
+          url,
+          headers: DATA_CONNECT_CONFIG_HEADERS,
+          data,
+        };
+        const resp = await this.httpClient.send(request);
+        return Promise.resolve({
+          data: resp.data.data as GraphqlResponse,
+        });
+      })
+      .then((resp) => {
+        return resp;
+      })
+      .catch((err) => {
+        throw this.toFirebaseError(err);
+      });
+  }
+
+  private async getUrl(host: string, locationId: string, serviceId: string, endpointId: string): Promise<string> {
+    return this.getProjectId()
+      .then((projectId) => {
+        const urlParams = {
+          host,
+          projectId,
+          locationId,
+          serviceId,
+          endpointId
+        };
+        const baseUrl = utils.formatString(DATA_CONNECT_API_URL_FORMAT, urlParams);
+        return utils.formatString(baseUrl);
+      });
+  }
+
+  private getProjectId(): Promise<string> {
+    if (this.projectId) {
+      return Promise.resolve(this.projectId);
+    }
+    return utils.findProjectId(this.app)
+      .then((projectId) => {
+        if (!validator.isNonEmptyString(projectId)) {
+          throw new FirebaseDataConnectError(
+            'unknown-error',
+            'Failed to determine project ID. Initialize the '
+            + 'SDK with service account credentials or set project ID as an app option. '
+            + 'Alternatively, set the GOOGLE_CLOUD_PROJECT environment variable.');
+        }
+        this.projectId = projectId;
+        return projectId;
+      });
+  }
+
+  private toFirebaseError(err: RequestResponseError): PrefixedFirebaseError {
+    if (err instanceof PrefixedFirebaseError) {
+      return err;
+    }
+
+    const response = err.response;
+    if (!response.isJson()) {
+      return new FirebaseDataConnectError(
+        'unknown-error',
+        `Unexpected response with status: ${response.status} and body: ${response.text}`);
+    }
+
+    const error: Error = (response.data as ErrorResponse).error || {};
+    let code: DataConnectErrorCode = 'unknown-error';
+    if (error.status && error.status in DATA_CONNECT_ERROR_CODE_MAPPING) {
+      code = DATA_CONNECT_ERROR_CODE_MAPPING[error.status];
+    }
+    const message = error.message || `Unknown server error: ${response.text}`;
+    return new FirebaseDataConnectError(code, message);
+  }
+}
+
+interface ErrorResponse {
+  error?: Error;
+}
+
+interface Error {
```

Comment:
Is Error not an existing native type that comes with TS? Maybe ServerError or ResponseError? I guess the second one looks kind of silly as a child of ErrorResponse.

Reply:
I don't think there is a specific reason to call it `Error`. It's just how we have used it in all the previous service implementations. I will rename this to `ServerError`. Thanks!